File: src/test/java/org/hisp/dhis/expression/ProgramRuleVariableExpressionTest.java
package org.hisp.dhis.expression;

import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Tests that program rule variables are identified and collected.
 *
 * @author Jan Bernitt
 */
class ProgramRuleVariableExpressionTest {

    @Test
    void testProgramRuleVariableString() {
        assertEquals(Set.of("var1"), evaluate("d2:count('var1')"));
    }

    @Test
    void testProgramRuleVariableName() {
        assertEquals(Set.of("var1"), evaluate("#{var1}"));
        assertEquals(Set.of("var1"), evaluate("A{var1}"));
    }

    private static Set<String> evaluate(String expression) {
        return new Expression(expression).collectProgramRuleVariables();
    }
}

File: src/jvmTest/java/org/hisp/dhis/lib/expression/ProgramRuleVariableExpressionTest.kt
package org.hisp.dhis.lib.expression

import org.hisp.dhis.lib.expression.Expression.Mode

import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Tests that program rule variables are identified and collected.
 *
 * @author Jan Bernitt
 */
internal class ProgramRuleVariableExpressionTest {
    @Test
    fun testProgramRuleVariableString() {
        assertEquals(setOf("var1"), evaluate("d2:count('var1')"))
    }

    @Test
    fun testProgramRuleVariableName() {
        assertEquals(setOf("var1"), evaluate("#{var1}"))
        assertEquals(setOf("var1"), evaluate("A{var1}"))
    }

    companion object {
        private fun evaluate(expression: String): Set<String> {
            return Expression(expression, Mode.RULE_ENGINE_ACTION).collectProgramRuleVariableNames()
        }
    }
}
